libs/fluid-elements/virtual-scroll-container/src/lib/virtual-scroll-state.ts

```diff
@@ -0,0 +1,291 @@
+   */
+  renderedItemsRange: FluidVirtualScrollRenderedItemRange = Object.seal({
+    first: -1,
+    last: -1,
+  });
+
+  /** Empty placeholder before the rendered items to correctly render the scrollbar */
+  startPlaceholderSize = 0;
+
+  /** Empty placeholder after the rendered items to correctly render the scrollbar */
+  endPlaceholderSize = 0;
+
+  /** Number of currently rendered start buffer items */
+  startItemsBufferSize = 0;
+
+  /** Number of currently rendered end buffer items */
+  endItemsBufferSize = 0;
+
+  /** Index of the item to observe for adjusting the start buffer */
+  bufferStartObservedItemIndex = -1;
+
+  /** Index of the item to observe for adjusting the end buffer */
+  bufferEndObservedItemIndex = -1;
+
+  /** Total number of items in the virtual scroll panel */
+  set itemsCount(value: number) {
+    if (value < 0) {
+      console.warn('Invalid item count: Item count should not be negative.');
+      return;
+    }
+    // Reset virtual scroll indices
+    this.reset();
+    this._itemsCount = value;
+  }
+  private _itemsCount = 0;
+
+  /** Number of items to render when initialising */
+  private _itemsBatchSize: number;
+
+  /** Number of initial batches rendered to fill up the whole container */
+  private _initialBatches = 0;
+
+  /** Number of items rendered above and below the fold */
+  private _itemsBufferSize: number;
+
+  /** Index of the first rendered item */
+  private _firstRenderedItemIndex = -1;
+
+  /** Index of the last rendered item */
+  private _lastRenderedItemIndex = -1;
+
+  constructor({
+    itemsBatchSize,
+    itemsBufferSize,
+  }: FluidVirtualScrollOptions = {}) {
+    this._itemsBatchSize = itemsBatchSize
+      ? itemsBatchSize
+      : VIRTUAL_SCROLL_BATCH_SIZE;
+    this._itemsBufferSize = itemsBufferSize
+      ? itemsBufferSize
+      : VIRTUAL_SCROLL_BUFFER_SIZE;
+  }
+
+  /** Initializes the array of items to render for the virtual scroll */
+  addInitialItemsBatch(): void {
+    if (this._itemsCount) {
+      this._appendItems(this._itemsBatchSize);
+      this._initialBatches += 1;
+    }
+  }
+
+  /** Adds new items to the start of the list, remove items from the end */
+  updateStartScrollBuffer(): void {
+    const prevRenderedItemsCount = this._renderedItemsCount();
+
+    // Add items to the start
+    this._prependItems(this._itemsBufferSize);
+
+    // Calculate start buffer size
+    this.startItemsBufferSize =
+      this._renderedItemsCount() - prevRenderedItemsCount;
+
+    // Remove items from the end depending on the amount of newly rendered items
+    this.renderedItemsRange.last -= this.startItemsBufferSize;
+    this._lastRenderedItemIndex = this.renderedItemsRange.last;
+
+    // Calculate number of items of the end buffer
+    this.endItemsBufferSize =
+      this._renderedItemsCount() -
+      this._initialBatches * this._itemsBatchSize -
+      this.startItemsBufferSize;
+
+    this._calculateObservedItemsIndices();
+  }
```

Comment:
A state should not have methods in my opinion. A state should be an object that holds a state not more.